## Dart'ta İstisnalar (Exceptions)
Dart'ta, istisnalar (exceptions) hata durumlarını işlemek ve hata yönetimini sağlamak için kullanılan mekanizmalardır. İstisnalar, beklenmedik durumlar veya hatalar oluştuğunda programın normal akışını kesintiye uğratır ve programın hatayı nasıl ele alacağını belirlemesine olanak tanır.

Dart'ta istisnaların temel mekanizması try-catch bloklarıdır. İşlem yapılacak kod try bloğu içine yerleştirilir ve olası hataların yakalanması ve işlenmesi için catch bloğu kullanılır.

İşte bir örnek:
```dart
void main() {
  try {
    // Potansiyel hata oluşabilecek kod
    int result = 10 ~/ 0; // Bölme işlemi sırasında hata oluşacak
    print('Sonuç: $result');
  } catch (e) {
    // Hata yakalandığında yapılacak işlemler
    print('Hata: $e');
  }
}

```
Yukarıdaki örnekte, bölme işleminde sıfıra bölme hatası (IntegerDivisionByZeroException) meydana gelir. Bu hata catch bloğunda yakalanır ve ekrana "Hata: [hata mesajı]" şeklinde bir mesaj yazdırılır.

catch bloğu, istisnaların türlerine göre de ayarlanabilir. Özel bir hata türünü yakalamak istiyorsanız, catch bloğunda o hatanın türünü belirtebilirsiniz.
```dart
void main() {
  try {
    // Potansiyel hata oluşabilecek kod
    int result = 10 ~/ 0; // Bölme işlemi sırasında hata oluşacak
    print('Sonuç: $result');
  } on IntegerDivisionByZeroException {
    // Sadece sıfıra bölme hatasını yakala
    print('Sıfıra bölme hatası!');
  } catch (e) {
    // Diğer hatalar için genel hata mesajı
    print('Hata: $e');
  }
}

```
Yukarıdaki örnekte, on anahtar kelimesiyle sadece IntegerDivisionByZeroException türündeki hatalar yakalanırken, diğer hatalar için genel bir catch bloğu kullanılır.

Yukarıdaki örnekte, finally bloğu, istisnaların oluşup yakalanması veya yakalanmaması durumuna bakılmaksızın her durumda çalıştırılacak olan kodu içerir. Bu blok, temizleme işlemleri, kaynakların serbest bırakılması veya sonlandırma işlemleri gibi önemli işlemleri gerçekleştirmek için kullanılabilir.
```dart
void main() {
  try {
    // Potansiyel hata oluşabilecek kod
    int result = 10 ~/ 0; // Bölme işlemi sırasında hata oluşacak
    print('Sonuç: $result');
  } catch (e) {
    // Hata yakalandığında yapılacak işlemler
    print('Hata: $e');
  } finally {
    // Her durumda çalışacak olan kod
    print('Sonlandırma işlemleri');
  }
}
```
Yukarıdaki örnekte, finally bloğu, her durumda "Sonlandırma işlemleri" metnini ekrana yazdıracaktır. Bu blok, hata yakalansa da yakalanmasa da çalıştırılacak kodları içerir.

Dart'ta, istisnalar programın hatayı ele almasını ve hatanın nasıl işleneceğini belirlemesini sağlayan önemli bir yapıdır. İstisnalar, hata ayıklama ve hata yönetimi süreçlerinde büyük ölçüde yardımcı olur ve uygulamaların daha güvenli ve kararlı çalışmasını sağlar.